include/nbl/asset/filters/CNormalMapToDerivativeFilter.h

```diff
@@ -0,0 +1,328 @@
+					CBasicImageFilterCommon::clip_region_functor_t clipFunctor(subresource, range, inFormat);
+
+					auto& inRegions = state->inImage->getRegions(state->inMipLevel);
+					CBasicImageFilterCommon::executePerRegion(state->inImage, decodeAndDivide, inRegions.begin(), inRegions.end(), clipFunctor);
+				}
+
+				{
+					auto getScratchPixel = [&](core::vector4di32_SIMD readBlockPos) -> CNormalMapToDerivativeFilterBase<Swizzle>::CNormalMapToDerivativeStateBase::decodeType*
+					{
+						const size_t scratchOffset = asset::IImage::SBufferCopy::getLocalByteOffset(core::vector3du32_SIMD(readBlockPos.x, readBlockPos.y, readBlockPos.z, 0), scratchByteStrides); // TODO
+						return reinterpret_cast<CNormalMapToDerivativeFilterBase<Swizzle>::CNormalMapToDerivativeStateBase::decodeType*>(reinterpret_cast<uint8_t*>(state->scratchMemory) + scratchOffset);
+					};
+
+					auto computeMaxAbs = [&](core::vectorSIMDi32 readBlockPos) -> void
+					{
+						auto* current = getScratchPixel(readBlockPos);
+						auto& [x, y] = std::make_tuple(*current, *(current + 1));
+
+						auto absoluteX = core::abs(x);
+						auto absoluteY = core::abs(y);
+
+						if (xMaxDecodeAbsValue < absoluteX)
+							xMaxDecodeAbsValue = absoluteX;
+
+						if (yMaxDecodeAbsValue < absoluteY)
+							yMaxDecodeAbsValue = absoluteY;
+					};
+
+					{
+						core::vector3du32_SIMD localCoord;
+						for (auto& z = localCoord[2] = 0u; z < state->extent.depth; ++z)
+							for (auto& y = localCoord[1] = 0u; y < state->extent.height; ++y)
+								for (auto& x = localCoord[0] = 0u; x < state->extent.width; ++x)
+									computeMaxAbs(core::vectorSIMDu32(x, y, z));
+					}
+
+					auto normalizeScratch = [&](bool isSigned)
+					{
+						core::vector3du32_SIMD localCoord;
+						for (auto& z = localCoord[2] = 0u; z < state->extent.depth; ++z)
+							for (auto& y = localCoord[1] = 0u; y < state->extent.height; ++y)
+								for (auto& x = localCoord[0] = 0u; x < state->extent.width; ++x)
+								{
+									const size_t scratchOffset = asset::IImage::SBufferCopy::getLocalByteOffset(localCoord, scratchByteStrides);
+									auto* entryScratchAdress = reinterpret_cast<CNormalMapToDerivativeFilterBase<Swizzle>::CNormalMapToDerivativeStateBase::decodeType*>(reinterpret_cast<uint8_t*>(state->scratchMemory) + scratchOffset);
+
+									if (isSigned)
+										for (uint8_t channel = 0; channel < CNormalMapToDerivativeFilterBase<Swizzle>::CNormalMapToDerivativeStateBase::forcedScratchChannelAmount; ++channel)
+											entryScratchAdress[channel] = entryScratchAdress[channel] / decodeAbsValuesOffset[channel];
+									else
+										for (uint8_t channel = 0; channel < CNormalMapToDerivativeFilterBase<Swizzle>::CNormalMapToDerivativeStateBase::forcedScratchChannelAmount; ++channel)
+											entryScratchAdress[channel] = entryScratchAdress[channel] * 0.5f / decodeAbsValuesOffset[channel] + 0.5f;
```

Comment:
the derivative map from heightmap filter needs a similar normalization (may need to inherit and slightly override the blit filter)

Reply:
btw does the blit filter have a temporary scratch for a whole decoded image like this filter?

Reply:
> btw does the blit filter have a temporary scratch for a whole decoded image like this filter?

it appears it does

```C++
// this will be called by derived classes because it doesn't account for all scratch needed, just the stuff for coverage adjustment
		static inline uint32_t getRequiredScratchByteSize(	typename CStateBase::E_ALPHA_SEMANTIC alphaSemantic=CStateBase::EAS_NONE_OR_PREMULTIPLIED,
															const core::vectorSIMDu32& outExtentLayerCount=core::vectorSIMDu32(0,0,0,0))
		{
			uint32_t retval = 0u;
			// 
			if (alphaSemantic==CStateBase::EAS_REFERENCE_OR_COVERAGE)
			{
				// no mul by channel count because we're only after alpha
				retval += outExtentLayerCount.x*outExtentLayerCount.y*outExtentLayerCount.z;
			}
			return retval*sizeof(typename value_type);
		}
```

Reply:
> 
> 
> btw does the blit filter have a temporary scratch for a whole decoded image like this filter?

then you could extend blit to provide the stastic on the max and min values of all channels instead?

Reply:
and do the normalization on that memory anyway

Reply:
sure